src/pinnable_mapped_file.cpp

```diff
@@ -265,36 +264,39 @@ void pinnable_mapped_file::revert_to_private_mode() {
 }
 
 // returns the number of pages flushed to disk
-size_t pinnable_mapped_file::check_memory_and_flush_if_needed() {
+std::optional<std::pair<size_t, size_t>> pinnable_mapped_file::check_memory_and_flush_if_needed() {
    size_t written_pages {0};
-#if 0
    if (_non_file_mapped_mapping || _sharable || !_writable)
-      return written_pages;
+      return {};
 
    // we are in `copy_on_write` mode.
    static time_t check_time = 0;
-   constexpr int check_interval = 60; // seconds
-   constexpr size_t one_gb = 1ull << 30;
+   constexpr int check_interval = 30; // seconds
 
    const time_t current_time = time(NULL);
    if(current_time >= check_time) {
       check_time = current_time + check_interval;
 
-      size_t avail_ram_gb = (get_avphys_pages() * sysconf(_SC_PAGESIZE)) / one_gb;
-      if (avail_ram_gb <= 2) {
-         auto [src, sz] = get_region_to_save();
-         pagemap_accessor pagemap;
-         size_t offset = 0;
-         while(offset != sz && written_pages < (one_gb / sysconf(_SC_PAGESIZE))) {
-            size_t copy_size = std::min(_db_size_copy_increment,  sz - offset);
-            if (!pagemap.update_file_from_region({ src + offset, copy_size }, _file_mapping, offset, false, written_pages))
-               break;
-            offset += copy_size;
-         }
+      auto oom_score = pagemap_accessor::read_oom_score();
+      if (oom_score) {
+         if (*oom_score >= 980) {
+            // linux returned a high out-of-memory (oom) score for the current process, indicating a high 
+            // probablility that the process will be killed soon (The valid range is from 0 to 1000.
+            // The higher the value is, the higher the probability is that the process will be killed).
```

Comment:
afaict that's not quite how it's documented as behaving, but rather just that a "A higher score means that the process is more likely to be selected by the OOM-killer." The absolute value doesn't matter, only its relative value compared to all the other processes' value. A process could still be killed due to system wide memory pressure at values less than 1000.

While in one limited test 980 may have made sense, it's not clear that value is appropriate if there are multiple nodeos running on the same box, as a simple example. Or, what happens if a lot of memory is consumed in tmpfs? Or, how do containers factor in to this?

I'd like to see more research on appropriate ways to detect system wide memory pressure (which is really what we want in this case).  For example,
https://www.kernel.org/doc/html/latest/accounting/psi.html
kinda-sorta sounds more like what we're looking for, but I'm not entirely sure.

Reply:
in the cases you mention (multiple nodeos,a lot of memory is consumed in tmpfs),  if users select `mapped_private`, there is a high likelihood that the nodeos will be killed by the OOM-killer.

This change just allows nodeos to decrease its OOM score when it gets very high.

Do I claim this PR catches all cases? No. 

Would some additional research allow us to do a better job at predicting when this memory relief is needed? Probably.

**Is it better to catch some at least some clear cases of OOM danger and avoid crashing if we can. Yes IMHO.**

Reply:
> This change just allows nodeos to decrease its OOM score when it gets very high.

I would be interested in doing more research, but my understanding right now is that a high absolute `oom_score` isn't really guaranteed to mean anything. It's only that score relative to other processes that mean something. And even then, it only means that the process with the highest score will be first on the chopping block next time the OOM killer takes action (whenever that is). These are all important distinctions to what has been written the comment.

Leaning on the absolute value of `oom_score` to determine when the process is getting close to being killed by OOM, appears to be leaning on implementation specific behavior. For example, consider if Linux changed in the next release so that the process using the most memory has a fixed `oom_score` of 1000, then the process with the second most fixed 999, etc etc. That still is completely valid by the rules of: "A higher score means that the process is more likely to be selected by the OOM-killer. The basis for this score is the amount of memory used by the process"

Reply:
Manually testing this more makes me more convinced `oom_score` isn't a good metric. For example, I created a 32GB VM, and set `chain-state-db-size-mb = 24576`. Once nodeos gets to running, I would get a flush about every 85000 blocks. Granted, this is nearly 12 hours of real time (but maybe not really real time since we don't know how speculative trxs dirty things), but it was every 60-90 seconds during a replay. When the flush was triggered nodeos only had 15-16GB resident and,
```
chainbase: oom = 984, flushed 822054 state pages to disk (in 16065 ms) to decrease memory pressure
chainbase: oom score after flush: 675
chainbase: The system is running low in memory for running nodeos. If this is not a temporary condition, consider increasing the available RAM
```
meaning just ~3GB was dirty.

There was clearly no memory pressure at all: _literally half_ of the VM's memory was completely unused. Yet, we're telling the user the system is running low on memory.